lib/commands/timeout.js

```diff
@@ -1,25 +1,46 @@
 import logger from '../logger';
-
+import { BaseDriver } from 'appium-base-driver';
+import { util } from 'appium-support';
 
 let commands = {}, helpers = {}, extensions = {};
 
-commands.timeouts = async function (type, duration) {
-  let ms = this.parseTimeoutArgument(duration);
-  switch (type) {
-    case 'command':
-      this.setNewCommandTimeout(ms);
-      break;
-    case 'implicit':
-      this.setImplicitWait(ms);
-      break;
-    case 'page load':
-      this.setPageLoadTimeout(ms);
-      break;
-    case 'script':
-      this.setAsyncScriptTimeout(ms);
-      break;
-    default:
-      throw new Error(`Invalid timeout '${type}'`);
+// Arguments will be: [{"protocol":"W3C","implicit":30000}, "1dcfe021-8fc8-49bd-8dac-e986d3091b97", ...]
+// eslint-disable-next-line no-unused-vars
+commands.timeouts = async function (timeoutsObj, ...args) {
+  if (timeoutsObj.protocol === BaseDriver.DRIVER_PROTOCOL.W3C) {
+    const {script, pageLoad, implicit} = timeoutsObj;
+    logger.debug(`script: ${script}, pageLoad: ${pageLoad}, implicit: ${implicit}`);
+
+    if (util.hasValue(script)) {
+      this.asyncScriptTimeout(script);
+    }
+
+    if (util.hasValue(pageLoad)) {
+      this.setPageLoadTimeout(this.parseTimeoutArgument(pageLoad));
+    }
+
+    if (util.hasValue(implicit)) {
+      this.implicitWait(implicit);
+    }
+  } else {
+    const {type, ms} = timeoutsObj;
+
+    switch (type) {
+      case 'command':
+        this.newCommandTimeout(ms);
+        break;
+      case 'implicit':
+        this.implicitWait(ms);
+        break;
+      case 'page load':
+        this.setPageLoadTimeout(this.parseTimeoutArgument(ms));
```

Comment:
shall we apply  `parseTimeoutArgument` to all method calls like it was in the previous implementation or to this one only?

Reply:
oops, I forgot to wrap this line...